ompi/mca/coll/base/coll_base_util.c

```diff
@@ -297,11 +298,232 @@ int ompi_coll_base_retain_datatypes_w( ompi_request_t *req,
     return OMPI_SUCCESS;
 }
 
-static void nbc_req_cons(ompi_coll_base_nbc_request_t *req) {
+static void nbc_req_cons(ompi_coll_base_nbc_request_t *req)
+{
     req->cb.req_complete_cb = NULL;
     req->req_complete_cb_data = NULL;
     req->data.objs.objs[0] = NULL;
     req->data.objs.objs[1] = NULL;
 }
 
 OBJ_CLASS_INSTANCE(ompi_coll_base_nbc_request_t, ompi_request_t, nbc_req_cons, NULL);
+
+/* File reading functions */
+static void skiptonewline (FILE *fptr, int *fileline)
+{
+    char val;
+    int rc;
+
+    do {
+        rc = fread(&val, 1, 1, fptr);
+        if (0 == rc) {
+            return;
+        }
+        if ('\n' == val) {
+            (*fileline)++;
+            return;
+        }
+    } while (1);
+}
+
+int ompi_coll_base_file_getnext_long(FILE *fptr, int *fileline, long* val)
+{
+    char trash;
+    int rc;
+
+    do {
+        rc = fscanf(fptr, "%li", val);
+        if (rc == EOF) {
+            return -1;
+        }
+        if (1 == rc) {
+            return 0;
+        }
+        /* in all other cases, skip to the end of the token */
+        rc = fread(&trash, sizeof(char), 1, fptr);
+        if (rc == EOF) {
+            return -1;
+        }
+        if ('\n' == trash) (*fileline)++;
+        if ('#' == trash) {
+            skiptonewline (fptr, fileline);
+        }
+    } while (1);
+}
+
+int ompi_coll_base_file_getnext_string(FILE *fptr, int *fileline, char** val)
+{
+    char trash, token[32];
+    int rc;
+
+    *val = NULL;  /* security in case we fail */
+    do {
+        rc = fscanf(fptr, "%32s", token);
+        if (rc == EOF) {
+            return -1;
+        }
+        if (1 == rc) {
+            if( '#' == token[0] ) {
+                skiptonewline(fptr, fileline);
+                continue;
+            }
+            *val = (char*)malloc(strlen(token) + 1);
+            strcpy(*val, token);
+            return 0;
+        }
+        /* in all other cases, skip to the end of the token */
+        rc = fread(&trash, sizeof(char), 1, fptr);
+        if (rc == EOF) {
+            return -1;
+        }
+        if ('\n' == trash) (*fileline)++;
+        if ('#' == trash) {
+            skiptonewline (fptr, fileline);
+        }
+    } while (1);
+}
+
+int ompi_coll_base_file_getnext_size_t(FILE *fptr, int *fileline, size_t* val)
+{
+    char trash;
+    int rc;
+
+    do {
+        rc = fscanf(fptr, "%" PRIsize_t, val);
+        if (rc == EOF) {
+            return -1;
+        }
+        if (1 == rc) {
+            return 0;
+        }
+        /* in all other cases, skip to the end of the token */
+        rc = fread(&trash, sizeof(char), 1, fptr);
+        if (rc == EOF) {
+            return -1;
+        }
+        if ('\n' == trash) (*fileline)++;
+        if ('#' == trash) {
+            skiptonewline (fptr, fileline);
+        }
+    } while (1);
+}
+
+int mca_coll_base_name_to_colltype(const char* name)
```

Comment:
This name to COLLTYPE conversion helper function with all the if and positional character logic looks confusing and hard to maintain for me.
I suggest using a conversion table instead.

Reply:
I would gladly review your PR.

Reply:
See https://github.com/bosilca/ompi/pull/24

Reply:
Sorry, a linear search is not a good replacement for an optimized search on a fixed space.

Reply:
Not sure if this is performance critical as my understanding is, that it is used once for parsing the dynamic rules file during init or is used in some verbose log output in case of problems.

But if you bother about performance in this case I have an optimized c99 version as well... (https://github.com/bosilca/ompi/compare/coll/han...cniethammer:colltype_translation_table_c99)
Here a quick benchmark (compiled with gcc -O2 and run on my Intel i5) iterating contiguously over all possible values with the pull request (cn) the current (gb) and the c99/optimized (cn_c99) implementation showing average user/CPU time per call and total iterations used for averaging:
```
-----------------------------------------------------------------------------------
Benchmark                                         Time             CPU   Iterations
-----------------------------------------------------------------------------------
BM_cn_mca_coll_base_name_to_colltype           31.6 ns         31.5 ns     23909947
BM_cn_mca_coll_base_colltype_to_str            48.3 ns         48.2 ns     23531703
BM_gb_mca_coll_base_name_to_colltype           14.0 ns         14.0 ns     63012206
BM_gb_mca_coll_base_colltype_to_str            4.44 ns         4.44 ns    118442753
BM_cn_c99_mca_coll_base_name_to_colltype       9.59 ns         9.58 ns     70586190
BM_cn_c99_mca_coll_base_colltype_to_str        3.25 ns         3.25 ns    219956199
```

I would go for the simpler, smaller code and if c99 is allowed also the optimized table version.

Reply:
The array approach for for c99 is indeed simpler, I will include it. For the linear search, I don't know how you did your benchmarking your results do not match [mine](https://gist.github.com/bosilca/f3bd08167acfb76f0a7d558b17ed8931). My approach does a dichotomy search instead of a linear one, I can hardly see in which cases it can be slower. For now I'll stick with my version for the string conversion.

Reply:
I used the google benchmark framework and cycled through all possible values to search: https://gist.github.com/cniethammer/e43990c39dd4a7669e7f5ace741a7f11
Not sure of the origin for the linear search being faster.
Looking at the produced assembler I see in the dichotomy search code strcmp inlined a lot producing a lot of code/instructions, in contrast the linear search loop just sticks with a call to strcmp: https://godbolt.org/z/P3nbxa

Reply:
Funny, the -O2 optimizations have not collapsed the switch, but -O3 does. Anyway, let's leave this as an academic exercise, the mixed code we have now in will be efficient for both operations in all cases.

Reply:
I did a quick sanity check on the return values. Both implementations produce the same result - all good to go for me.

If we find by accident spare time during an Open MPI telco I am happy to dive deeper and discuss performance impacts for HAN - but for now I will shift focus to new topics. ;)